### PR TITLE
general: bump inspire-schemas to version ~26.0

### PR DESCRIPTION
**Pull request by jacquerie**
Closes #1883
Closes #1886 
Closes #1891 
Closes #1894 
Closes #1907 
Closes #1910 
Closes #1914 
Closes #1917 
Closes #1933 
Closes #1937

This PR will contain an atomic* commit that bumps all the code to support `inspire-schemas~=25.0`. For now, it only fixes (almost all) the DoJSON rules so that they satisfy the spec.

Lessons learned: don't do it like this. `inspire-schemas` has major versions, and an update between major versions already requires some effort. ~~13~~ ~~17~~ ~~18~~ 21 times "some effort" is "too much effort". So we should have made several smaller PRs that handled 4->5, 5->6... (it's too late for that).

Moreover, long-lived branches are too hard to integrate, and generate so many PRs that it's hard to keep track who did what (look at that list above). So we shouldn't do those either. As much as possible we should develop on trunk.

\* In the same sense in which a bomb can be called "atomic".

### CODE REVIEW

File: inspirehep/dojson/common/base.py [hunk 1]
```diff
@@ -54,20 +56,34 @@ def _get_source(value):
 
         return force_single_element(sources_without_inspire_uid)
 
+    def _get_orcid(value):
+        orcids = force_force_list(value.get('a'))
```
Comment (michamos) [inline inspirehep/dojson/common/base.py]: `orcids` -> `sources`?

File: inspirehep/dojson/common/base.py [hunk 2]
```diff
@@ -105,9 +121,7 @@ def control_number2marc(self, key, value):
 @journals.over('spires_sysnos', '^970..')
 @hepnames.over('spires_sysnos', '^970..')
 @jobs.over('spires_sysnos', '^970..')
-@utils.ignore_value
 def spires_sysnos(self, key, value):
-    """Old SPIRES number and new_recid from 970."""
     external_system_numbers = self.get('external_system_numbers', [])
```
Comment (michamos) [inline inspirehep/dojson/common/base.py]: Should we use this opportunity to unify all schemas by using `external_system_identifiers` like in hep instead?

File: inspirehep/dojson/hep/fields/bd0xx.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+RE_LANGUAGE = re.compile('\/| or | and |,|=|\s+')
+
+
+@hep.over('isbns', '^020..')
+@utils.for_each_value
+def isbns(self, key, value):
+    def _get_medium(value):
+        def _normalize(medium):
+            schema = load_schema('hep')
+            valid_mediums = schema['properties']['isbns']['items']['properties']['medium']['enum']
```
Comment (michamos) [inline inspirehep/dojson/hep/fields/bd0xx.py]: [`valid_media`](http://grammarist.com/usage/media-mediums/)?

File: inspirehep/dojson/hep/fields/bd0xx.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+            return ''
+
+        medium = force_single_element(value.get('b', ''))
+        normalized_medium = _normalize(medium)
+
+        return normalized_medium
+
+    def _get_isbn(value):
+        def _normalize(isbn):
+            return isbn.lower().replace('-', '')
```
Comment (michamos) [inline inspirehep/dojson/hep/fields/bd0xx.py]: Shouldn't it be `upper()` instead? The schema contains capital `X`.

File: inspirehep/dojson/hep/fields/bd0xx.py [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+
+
+@hep.over('dois', '^0247.')
+def dois(self, key, value):
+    """Populate the ``dois`` key.
+
+    Also populates the ``persistent_identifiers`` key through side effects.
+    """
+    def _get_first_non_curator_source(sources):
+        sources_without_curator = [el for el in sources if el.upper != 'CURATOR']
```
Comment (michamos) [inline inspirehep/dojson/hep/fields/bd0xx.py]: `el.upper` -> `el.upper()`?

File: inspirehep/dojson/hep/fields/bd0xx.py [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+        return force_single_element(sources_without_curator)
+
+    def _is_doi(id_, type_):
+        return (not type_ or type_.upper() == 'DOI') and is_doi(id_)
+
+    def _is_handle(id_, type_):
+        return (not type_ or type_.upper() in ('DOI', 'HDL')) and is_handle(id_)
```
Comment (michamos) [inline inspirehep/dojson/hep/fields/bd0xx.py]: Can you explain why you something having `type_ == 'DOI'` should be considered as a handle?

File: inspirehep/dojson/hep/fields/bd0xx.py [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+        source = _get_first_non_curator_source(sources)
+
+        if _is_doi(id_, schema):
+            dois.append({
+                'material': material,
+                'source': source,
+                'value': normalize_doi(id_),
+            })
+        else:
+            schema = 'HDL' if _is_handle(id_, schema) else schema
```
Comment (michamos) [inline inspirehep/dojson/hep/fields/bd0xx.py]: * If I understand this correctly, `0247_2:DOI` with invalid DOI gets put into `persistent_identifiers`, which fails because `schema` is an [enum](https://github.com/inspirehep/inspire-schemas/blob/4d88e8d128a3ea0d7465581fd72f3e4965f6b50f/inspire_schemas/records/hep.json#L598). Is that what should happen?
* Don't you similarly want to check for `URN`?

File: inspirehep/dojson/hep/fields/bd0xx.py [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+@hep.over('texkeys', '^035..')
+def texkeys(self, key, value):
+    """Populate the ``texkeys`` key.
+
+    Also populates the ``external_system_identifiers`` key through side effects.
+    """
+    def _is_texkey(id_, schema):
+        return id_ and schema in ('INSPIRETeX', 'SPIRESTeX')
+
+    def _is_other_texkey(other_id, schema):
```
Comment (michamos) [inline inspirehep/dojson/hep/fields/bd0xx.py]: `_is_other_texkey` == `_is_texkey`, or am I missing something?

File: inspirehep/dojson/hep/fields/bd1xx.py [hunk 9]
```diff
@@ -147,47 +148,29 @@ def _get_record(value):
...
-                    contributor_roles.append(
-                        {
-                            'schema': 'CRediT',
-                            'value': 'Project administration'
-                        }
-                    )
-            return contributor_roles
+        def _get_curated_relation(value):
+            if key.startswith('100'):
+                return value.get('y') == '1'
```
Comment (michamos) [inline inspirehep/dojson/hep/fields/bd1xx.py]: * You can can also get the value `2` here, according to Twiki.
* Why only for `100`?

File: inspirehep/dojson/hep/fields/bd0xx.py (hunk 5 above)
Comment (jacquerie) [inline inspirehep/dojson/hep/fields/bd0xx.py]: Whoops! This also means that we're missing a test, because this should have been caught there.

File: inspirehep/dojson/hep/fields/bd5xx.py [hunk 10]
```diff
@@ -64,11 +64,11 @@ def public_notes2marc(self, key, value):
...
-    """Hidden notes."""
-    def _hidden_notes(value):
-        def _hidden_note(value, a=None):
+@hep.over('_private_notes', '^595..')
+def private_notes(self, key, value):
+    """Private notes."""
+    def _private_notes(value):
+        def _private_note(value, a=None):
             source = value.get('9')
             if value.get('c'):
```
Comment (michamos) [inline inspirehep/dojson/hep/fields/bd5xx.py]: `595__c` can also contain information about HAL export (`HAL`/`NOT HAL`)

File: inspirehep/dojson/hep/fields/bd5xx.py [hunk 11]
```diff
@@ -64,11 +64,11 @@ def public_notes2marc(self, key, value):
     }
 
 
-@hep.over('hidden_notes', '^595..')
-def hidden_notes(self, key, value):
-    """Hidden notes."""
-    def _hidden_notes(value):
-        def _hidden_note(value, a=None):
+@hep.over('_private_notes', '^595..')
```
Comment (michamos) [inline inspirehep/dojson/hep/fields/bd5xx.py]: `595_D` should have `source` = `DESY`

File: inspirehep/dojson/hep/fields/bd5xx.py (hunk 10 above)
Comment (michamos) [inline inspirehep/dojson/hep/fields/bd5xx.py]: @kaplun, should we add something like `_export_to` in the schema, which is enum ['CDS', 'HAL',...] to tag records to export, or do you want to do this differently?

File: inspirehep/dojson/hep/fields/bd0xx.py (hunk 6 above)
Comment (kaplun) [inline inspirehep/dojson/hep/fields/bd0xx.py]: Because it is the reality on legacy. I guess this could be simplified by cleaning up legacy. Ha! It's only 30 records:
https://inspirehep.net/search?ln=en&ln=en&p=0247%3ADOI+-0247_a%3A10.*&of=hb&action_search=Search&sf=earliestdate&so=d&rm=&rg=25&sc=0&wl=0

File: inspirehep/dojson/hep/fields/bd0xx.py (hunk 4 above)
Comment (jacquerie) [inline inspirehep/dojson/hep/fields/bd0xx.py]: Yup. That's another test that was missing.

File: inspirehep/dojson/hep/fields/bd1xx.py (hunk 9 above)
Comment (kaplun) [inline inspirehep/dojson/hep/fields/bd1xx.py]: The TWIKI is wrong. It's either 0 or 1.
https://github.com/inspirehep/inspire/blob/master/bibformat/format_elements/bfe_INSPIRE_enhanced_marcxml.py#L192

Comment (kaplun) [inline inspirehep/dojson/hep/fields/bd1xx.py]: Just updated the TWiki

File: inspirehep/dojson/hep/fields/bd5xx.py (hunk 10 above)
Comment (kaplun) [inline inspirehep/dojson/hep/fields/bd5xx.py]: I think this is a very good idea. Indeed we are having these usecases. Additional one is: "candidate to be exported to". (somehow this was always implemented with special collections but to make them explicit in the data model allows to build more straightforward workflows.

File: inspirehep/dojson/hep/fields/bd0xx.py (hunk 6 above)
Comment (michamos) [inline inspirehep/dojson/hep/fields/bd0xx.py]: cleaned

File: inspirehep/dojson/hep/fields/bd70x75x.py [hunk 12]
```diff
@@ -66,19 +62,64 @@ def _get_affiliations(value):
...
         return result
 
+    def _get_ids(value):
+        def _is_jacow(j_value):
+            return j_value.upper().startswith('JACOW-')
+
+        def _is_orcid(j_value):
+            return j_value.upper().startswith('ORCID:') and len(j_value) > 6
+
+        def _is_naked_orcid(j_value):
```
Comment (michamos) [inline inspirehep/dojson/hep/fields/bd70x75x.py]: why not [idutils.is_orcid](http://idutils.readthedocs.io/en/latest/#idutils.is_orcid)?

File: inspirehep/dojson/hep/fields/bd9xx.py [hunk 13]
```diff
@@ -29,22 +29,150 @@
...
+@hep2marc.over('980', '^withdrawn$')
+@utils.for_each_value
+def withdrawn2marc(self, key, value):
+    if value:
+        return {'a': 'Withdrawn'}
+
+
+@hep2marc.over('980', '^publication_type$')
+@utils.for_each_value
+def publication_type2marc(self, key, value):
```
Comment (michamos) [inline inspirehep/dojson/hep/fields/bd9xx.py]: need to be normalized for legacy

File: inspirehep/dojson/hep/fields/bd9xx.py [hunk 14]
```diff
@@ -29,22 +29,150 @@
...
+
+@hep2marc.over('980', '^special_collections$')
+@utils.for_each_value
+def special_collections2marc(self, key, value):
+    return {'a': value}
+
+
+@hep2marc.over('980', '^document_type$')
+@utils.for_each_value
+def document_type2marc(self, key, value):
```
Comment (michamos) [inline inspirehep/dojson/hep/fields/bd9xx.py]: need to be normalized for legacy

File: inspirehep/dojson/hep/model.py [hunk 15]
```diff
@@ -24,64 +24,64 @@
...
 
 from ..model import FilterOverdo, add_schema, clean_record
-from ..utils import (
-    classify_field,
-    force_single_element,
-    get_record_ref,
-)
-
-
-def add_book_info(record, blob):
```
Comment (michamos) [inline inspirehep/dojson/hep/model.py]: this should happen in general for `773__0` -> `publication_info.parent_record`

File: inspirehep/dojson/hep/model.py [hunk 16]
```diff
@@ -24,64 +24,64 @@
...
-                except (ValueError, TypeError):
-                    pass
+from ..utils import classify_field
+
+
+def add_arxiv_categories(record, blob):
+    if not record.get('arxiv_eprints') or not blob.get('65017'):
+        return record
+
+    for arxiv_category in force_force_list(get_value(blob, '65017.a')):
```
Comment (michamos) [inline inspirehep/dojson/hep/model.py]: you should filter on `650172:arXiv`

File: tests/unit/dojson/test_dojson_common_base.py [hunk 17]
```diff
@@ -215,62 +213,12 @@ def test_new_record_from_970__d():
...
-
-    expected = [
-        {'a': 'HEP'},
-    ]
-    result = hep2marc.do(result)
-
-    assert expected == result['980']
-
-
-def test_deleted_from_980__c():
```
Comment (michamos) [inline tests/unit/dojson/test_dojson_common_base.py]: why is deleted deleted?

File: tests/unit/dojson/test_dojson_hep_bd0xx.py [hunk 18]
```diff
@@ -140,9 +139,7 @@ def test_isbns_from_020__a_b_normalizes_electronic():
 
     expected = [
         {
-            'value': '978-3-319-00626-0',
-            'medium': 'online',
```
Comment (michamos) [inline tests/unit/dojson/test_dojson_hep_bd0xx.py]: this line should be preserved

File: tests/unit/dojson/test_dojson_hep_bd0xx.py [hunk 19]
```diff
@@ -152,8 +149,7 @@ def test_isbns_from_020__a_b_normalizes_electronic():
 
     expected = [
         {
-            'a': '978-3-319-00626-0',
-            'b': 'online',
```
Comment (michamos) [inline tests/unit/dojson/test_dojson_hep_bd0xx.py]: same

File: tests/unit/dojson/test_dojson_hep_bd0xx.py [hunk 20]
```diff
@@ -507,49 +465,56 @@ def test_texkeys_from_multiple_035__a_z_9():
...
-    assert expected == result['external_system_identifiers']
-    assert expected == result['arxiv_eprints']
+    try:
+        validate(result.get('external_system_identifiers'), subschema)
+    except ValidationError:
+        assert True
 
 
-@pytest.mark.xfail(reason='wrong roundtrip')
 def test_external_system_numbers_from_035__a_d_h_m_9():
```
Comment (michamos) [inline tests/unit/dojson/test_dojson_hep_bd0xx.py]: `numbers` -> `identifiers`

File: tests/unit/dojson/test_dojson_hep_bd0xx.py [hunk 21]
```diff
@@ -717,51 +677,67 @@ def test_report_numbers_hidden_from_037__z():
     assert expected == result['037']
 
 
-@pytest.mark.xfail(reason='Schema 13')
-def test_report_numbers_source_from_037__z_9():
+def test_report_numbers_from_037__z_9():
```
Comment (michamos) [inline tests/unit/dojson/test_dojson_hep_bd0xx.py]: 037__9_a

File: tests/unit/dojson/test_dojson_hep_bd1xx.py [hunk 22]
```diff
@@ -133,7 +130,6 @@ def test_authors_from_100__a_u_w_y_and_700_a_u_w_x_y():
                     'value': 'Kyoto U.',
                 },
             ],
-            'curated_relation': True,
```
Comment (michamos) [inline tests/unit/dojson/test_dojson_hep_bd1xx.py]: this line should be here

File: tests/unit/dojson/test_dojson_hep_bd1xx.py [hunk 23]
```diff
@@ -217,7 +209,6 @@ def test_authors_from_100__a_e_w_y_and_700_a_e_w_y():
                     'value': 'B.A.Knyazev.2',
                 },
             ],
-            'curated_relation': False
```
Comment (michamos) [inline tests/unit/dojson/test_dojson_hep_bd1xx.py]: should be here

File: tests/unit/dojson/test_dojson_hep_bd1xx.py [hunk 24]
```diff
@@ -227,100 +218,20 @@ def test_authors_from_100__a_e_w_y_and_700_a_e_w_y():
 
     expected_100 = {
         'a': 'Vinokurov, Nikolay A.',
-        'e': 'ed.',
+        'e': ['ed.'],
     }
     expected_700 = [
         {
             'a': 'Knyazev, Boris A.',
-            'e': 'ed.'
```
Comment (michamos) [inline tests/unit/dojson/test_dojson_hep_bd1xx.py]: this whole piece should not disappear

File: tests/unit/dojson/test_dojson_hep_bd1xx.py (hunk 23 above)
Comment (kaplun) [inline tests/unit/dojson/test_dojson_hep_bd1xx.py]: This one might be left a part. When it's False or the field is not there is basically the same thing.

File: tests/unit/dojson/test_dojson_hep_bd1xx.py [hunk 25]
```diff
@@ -766,128 +688,28 @@ def test_authors_from_100__a_triple_u_w_x_y_triple_z_and_700__double_a_u_w_x_y_z
...
-            },
+    expected_100 = {
+        'a': 'Abe, K.',
+        'u': [
+            'Tokyo U., ICRR',
+            'Tokyo U.',
+            'Tokyo U., IPMU'
         ],
     }
-    result = hep2marc.do(result)
```
Comment (michamos) [inline tests/unit/dojson/test_dojson_hep_bd1xx.py]: why did this disappear?

File: tests/unit/dojson/test_dojson_hep_bd9xx.py [hunk 26]
```diff
@@ -221,21 +228,20 @@ def test_document_type_from_980__a():
...
-        </datafield>
-    '''
+    snippet = (
+        '<datafield tag="980" ind1=" " ind2=" ">'
+        '  <subfield code="a">Review</subfield>'
+        '</datafield>'
+    )  # record/1509993
 
-    expected = ['review']
+    expected = 'review'
```
Comment (michamos) [inline tests/unit/dojson/test_dojson_hep_bd9xx.py]: this is a schema bug that I am going to correct

File: tests/unit/dojson/test_dojson_hep_model.py [hunk 27]
```diff
@@ -1,75 +0,0 @@
-# -*- coding: utf-8 -*-
```
Comment (michamos) [inline tests/unit/dojson/test_dojson_hep_model.py]: why did you remove this?

**Comment by michamos** [comment]
There are additional tests in #1917 that you didn't take

File: tests/unit/dojson/test_dojson_hep_bd9xx.py (hunk 26 above)
Comment (michamos) [inline tests/unit/dojson/test_dojson_hep_bd9xx.py]: fixed in inspirehep/inspire-schemas#91, `publication_type` is now an array

File: inspirehep/dojson/hep/fields/bd5xx.py (hunk 11 above)
Comment (michamos) [inline inspirehep/dojson/hep/fields/bd5xx.py]: also `595_H` should have `source` = `HAL`

File: inspirehep/dojson/common/rules.py [hunk 28]
```diff
@@ -0,0 +1,396 @@
...
+    }
+
+
+@conferences.over('inspire_categories', '^65017')
+@experiments.over('inspire_categories', '^65017')
+@hep.over('inspire_categories', '^65017')
+@hepnames.over('inspire_categories', '^65017')
+@institutions.over('inspire_categories', '^65017')
+@jobs.over('inspire_categories', '^65017')
+def inspire_categories(self, key, value):
```
Comment (jmartinm) [inline inspirehep/dojson/common/rules.py]: Note that these are mostly arXiv categories (https://github.com/inspirehep/inspire-next/blob/master/inspirehep/dojson/current_marcxml_usage/hepnames.txt#L931) and the new field `arxiv_categories` (https://github.com/inspirehep/inspire-schemas/blob/master/inspire_schemas/records/authors.json#L54) should be used.

Comment (jmartinm) [inline inspirehep/dojson/common/rules.py]: Note that for `hepnames` these are mostly arXiv categories (https://github.com/inspirehep/inspire-next/blob/master/inspirehep/dojson/current_marcxml_usage/hepnames.txt#L931) and the new field `arxiv_categories` (https://github.com/inspirehep/inspire-schemas/blob/master/inspire_schemas/records/authors.json#L54) should be used.

**Comment by jmartinm** [comment]
We are missing a dojson conversion from `_fft` to `FFT` since the literature submission form writes into `_fft` and marcxml should be generated.

File: inspirehep/dojson/common/rules.py (hunk 28 above)
Comment (jacquerie) [inline inspirehep/dojson/common/rules.py]: Should be solved by https://github.com/inspirehep/inspire-next/pull/1918/commits/21f54124f57e5529fd9f9fc5f930bbbe24feaba4.
